### PR TITLE
Passing data via construct function and updating data structs for filled line graphs

### PR DESCRIPTION
**Pull request by tia-ylinen**
Previously we had a struct for a data point, I kept that and added a new one that consists of an array of those data points and name of that data set. Any further information about the data can be of course added to it but I didn't think of anything else for now?

### CODE REVIEW

File: Source/ImPlotWidgets/Public/ImPlotWidget.h [hunk 1]
```diff
@@ -10,24 +10,27 @@ struct FDataPoint
...
 class IMPLOTWIDGETS_API SImPlotWidget : public SCompoundWidget
 {
 private:
 	using Super = SCompoundWidget;
 public:
 	SLATE_BEGIN_ARGS(SImPlotWidget)
 	{}
-	SLATE_ARGUMENT(TArray<TArray<FDataPoint>>, DataArrays)
-	SLATE_ARGUMENT(TArray<float>, InData)
-	SLATE_ARGUMENT(TArray<char*>, DataSetNames)
```
Comment (willhinds) [inline Source/ImPlotWidgets/Public/ImPlotWidget.h:22]: We can delete the DataArrays and Data class members too.

File: Source/ImPlotWidgets/Public/ImPlotWidget.h [hunk 2]
```diff
@@ -10,24 +10,27 @@ struct FDataPoint
 	float Y = 0.0f;
 };
 
+struct FImPlotData
+{
+	char* Name;
```
Comment (willhinds) [inline Source/ImPlotWidgets/Public/ImPlotWidget.h]: I would change this from a char* to an FString now to fit in more with the Unreal way of doing things

**Review by willhinds — CHANGES_REQUESTED** [review]
Really good work! Only a couple of small changes and then it should be good to go.

File: Source/ImPlotWidgets/Private/HistogramWidget.cpp [hunk 3]
```diff
@@ -5,9 +5,10 @@
 #include <ImGui/Public/ImPlotImplementation.h>
 #include <Runtime/Core/Public/Algo/Sort.h>
 
-void SHistogramWidget::Construct(const FArguments& InArgs)
+void SHistogramWidget::Construct(const FArguments& InArgs, TArray<float> InData)
```
Comment (Josef-Rare) [inline Source/ImPlotWidgets/Private/HistogramWidget.cpp]: You should pass this TArray as a `const TArray&` Currently your code (nominally, as the compiler may be smart enough to optimise this away) is creating a TArray copy as part of the function call, then on line 11 you are copy assigning this array to your member variable. The result is two array constructions when you need one of those at most. There's another example of this pattern in SFilledGraphWidget::Construct.

Generally, unless it's a primitive data type (int, uintX, uint_X_t, float and so on) or occasionally a class / struct that has negligible construction overhead then you should pass params as const references.

Arrays and strings (which are specialised arrays) are expensive to construct as they have to make a dynamic (heap) allocation!

**Review by Josef-Rare — CHANGES_REQUESTED** [review]

File: Source/ImPlotWidgets/Private/HistogramWidget.cpp (hunk 3 above)
Comment (willhinds) [reply Source/ImPlotWidgets/Private/HistogramWidget.cpp]: Good spot @Josef-CL, would this have to be explicitly a const reference reference `const TArray<float>&&` to allow moving, or can you MoveTemp into a const reference fine? I can see lots of common scenarios where the data arrays would be constructed solely for use by the graph and so moving would be the best thing to do.

Comment (Josef-Rare) [reply Source/ImPlotWidgets/Private/HistogramWidget.cpp]: Yeah if you want that to be the API then you would have to make it an R-value (&&) param toa loow move semantics. But not const as you explicitly want to take ownership and allow move to happen.

**Review by willhinds — APPROVED** [review]

**Review by Josef-Rare — APPROVED** [review]

**Pull request merged**